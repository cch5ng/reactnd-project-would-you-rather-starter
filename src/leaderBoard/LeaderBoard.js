import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import '../App.css';
import { getArFromDict } from '../utilities/utilities';

export class LeaderBoard extends Component {
  constructor(props) {
    super(props);

    this.sortByAnswersCount = this.sortByAnswersCount.bind(this);
  }

  /*
   * @param {ar} [{user}, ...]
   * @return {ar} [{user}, ...]
   *
   * Given array of users, sorts by number of questions answered (max => min)
   */
  sortByAnswersCount(ar) {
    ar.sort((a, b) => {
      let bNumAnswers = Object.keys(b.answers).length;
      let aNumAnswers = Object.keys(a.answers).length;
      return bNumAnswers - aNumAnswers;
    });
    return ar;
  }

  render() {
    const {login, users} = this.props;
    let isLoggedIn;
    let usersAr;

    if (login && login.isLoggedIn) {
      isLoggedIn = login.isLoggedIn;
    }

    if (users && users.users) {
      usersAr = getArFromDict(users.users)
      usersAr = this.sortByAnswersCount(usersAr);
      console.log('usersAr: ' + JSON.stringify(usersAr));
    }

    return (
      <div className="">
        <h2>Leader Board</h2>

        {isLoggedIn && (
          usersAr.map(user => {
            let numQuestionsAnswered = Object.keys(user.answers).length;

            return (
              <div key={user.id}>
                <p>{user.name}</p>
                <p><img src={user.avatarURL} alt="user avatar" /></p>
                <p># Questions answered: {numQuestionsAnswered}</p>
                <p># Questions asked: {user.questions.length}</p>
              </div>
            )
          })
        )}

        {!isLoggedIn && (
          <div>Sorry, you need to log in to view this page.</div>
        )}

      </div>
    )
  }
}

function mapStateToProps({ login, users }) {
  return {
    login,
    users,
  }
}

export default connect(mapStateToProps)(LeaderBoard);
