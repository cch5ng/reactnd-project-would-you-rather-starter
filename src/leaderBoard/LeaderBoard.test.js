// LeaderBoard.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import { LeaderBoard } from '../leaderBoard/LeaderBoard.js';
import renderer from 'react-test-renderer';

test('LeaderBoard renders', () => {

  const div = document.createElement('div');

  ReactDOM.render(<LeaderBoard />, div);
  ReactDOM.unmountComponentAtNode(div);

  // snapshot test
  const tree = renderer
    .create(<LeaderBoard />,)
    .toJSON();

    expect(tree).toMatchSnapshot();
});
