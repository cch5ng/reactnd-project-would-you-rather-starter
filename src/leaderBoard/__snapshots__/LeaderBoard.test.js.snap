// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderBoard renders 1`] = `
<div
  className=""
>
  <h2>
    Leader Board
  </h2>
  <div>
    Sorry, you need to log in to view this page.
  </div>
</div>
`;
